lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+void StructSplitAnalyzer::makeSuggestions() {
+  if (!Eligibility) {
+    outs() << "Struct has too few fields or cold in profiling for splitting\n";
+    return;
+  }
+  unsigned dynamicThresholdUpdates = 0;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                  dbgs() << "Making suggestions for struct splitting\n");
+  auto BestPair = findMaxRemainCP();
+  double Threshold = CloseProximityRelations[BestPair.first][BestPair.second] /
+                     Params.ColdRatio;
+
+  while (FieldsToTransform.size()) {
+    if (FieldsToTransform.size() == 1) {
+      // Only has one field left, break the loop and create a subrecord for it
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                      dbgs() << "Only have one remaining field...\n");
+      break;
+    }
+    auto BestPair = findMaxRemainCP();
+    auto MaxCP = CloseProximityRelations[BestPair.first][BestPair.second];
+    if (MaxCP == 0)
+      break;
+    // If MaxCP is colder than the current threshold, adjust the threhold for
+    // several times until the current pair is hotter than threshold or reach
+    // the limit of lowering the bar
+    while (MaxCP < Threshold) {
+      if (dynamicThresholdUpdates++ > Params.MaxNumThresholdUpdates)
+        break;
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                      dbgs() << "Lowering threshold by a factor of "
+                             << Params.ColdRatio << "\n");
+      Threshold /= Params.ColdRatio;
+    }
+    if (MaxCP < Threshold)
+      break;
+    DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                    dbgs() << "CP value is: " << DEBUG_PRINT_COUNT(MaxCP)
+                           << " and threshold is: "
+                           << DEBUG_PRINT_COUNT(Threshold) << "\n");
+    auto *NewRecord = new SubRecordType;
+    SubRecords.push_back(NewRecord);
+    // Add the pair into the new sub record and remove from remaining fields
+    NewRecord->push_back(BestPair.first);
+    NewRecord->push_back(BestPair.second);
+    DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT, dbgs() << "Create a new record with F"
+                                             << BestPair.first + 1 << " and F"
+                                             << BestPair.second + 1 << "\n");
+    auto NewRecordSize =
+        FieldSizes[BestPair.first] + FieldSizes[BestPair.second];
+    FieldsToTransform.erase(BestPair.first);
+    FieldsToTransform.erase(BestPair.second);
+    // If the pair is enough to take the MAX_SIZE, no need to add more fields to
+    // the subrecord
+    if (NewRecordSize >= Params.MaxSize)
+      continue;
+    while (FieldsToTransform.size()) {
+      double MaxSum = 0;
+      FieldNumType MaxF = 0;
+      for (auto &Field : FieldsToTransform) {
+        // If the field is too large to fit in the new sub record, give up on
+        // this one
+        if (FieldSizes[Field] + NewRecordSize > Params.MaxSize)
+          continue;
+        auto Sum = 0;
+        for (auto &Fi : *NewRecord) {
```

Comment:
Will fix. I usually do it but this might be one of the several cases I thought I might need more statements but it turned out only one statement...